File: 2022/7/solution.py
import sys

root = {
        "parent": None,
        "name": "/",
        "children": {},
        "files": {},
        "size": 0
}

current = root

for line in sys.stdin.readlines():
    line = line.strip().split()
    if line[0] == "$":
        cmd = line[1]
        if cmd == "ls":
            continue

        arg = line[2]
        if arg == "..":
            current = current["parent"]
        elif arg == "/":
            continue
        else:
            current = current["children"][arg]
    else:
        size_or_type, name = line[0:]
        if size_or_type == "dir":
            current["children"][name] = {
                    "parent": current,
                    "name": name,
                    "children": {},
                    "files": {}
            }


        else:
            current["files"][name] = int(size_or_type)

pt1 = 0
alldirs = []

def size(fs):
    global pt1
    global alldirs

    children = [size(child) for child in fs["children"].values()]
    files = sum(fs["files"].values())

    me = sum(children) + files

    if me <= 100000:
        print(f"Adding {fs['name']} at size {me}")
        pt1 += me
    else:
        print(f"Not adding {fs['name']} at size {me}")

    alldirs.append(me) 
    return me


used = size(root)
print(f"Part 1: {pt1}")

total = 70000000
needed = 30000000
free = total - used
to_reclaim = needed - free

alldirs.sort()

for d in alldirs: 
    if d > to_reclaim:
        print(f"Part 2: {d}")
        break

File: 2022/7/solution.rs
use std::io::BufRead;
use std::io;
use std::collections::{HashMap, HashSet};

type TInput = Directory;

#[derive(Debug, Eq, PartialEq)]
struct Directory {
    name: String,
    files: HashMap<String, u32>,
    children: Vec<Directory>
}

fn sum(dir: &Directory) -> u32 {
    dir.files.values().sum::<u32>() + 
        dir.children.iter().map(sum).sum::<u32>()
}

fn part1(input: &TInput) -> u32 {
    0
}

fn part2(input: &TInput) -> u32 {
    0
}

fn main() -> io::Result<()> {
    let stdin = io::stdin();
    let reader = stdin.lock();
    let root = Directory { name: "/".to_string(), files: HashMap::new(), children: vec![] };
    let mut path: Vec<&Directory> = vec![];
    let mut current = root;

    for line in reader.lines() {
        let line = line.expect("Could not read stdin");
        if line == "$ cd /" {
            continue;
        }

        if line.starts_with("$") {
            let line = line.strip_prefix("$ ").unwrap();
            if let Some((_cd, arg)) = line.split_once(" ") {
                if arg == ".." {
                    current = *path.pop().unwrap();
                } else {
                    let next = current.children.iter_mut().filter(|c| c.name == arg).next().unwrap();
                    path.push(&current);
                    current = *next;
                }
            }
        } else {
            let (type_or_size, name) = line.split_once(" ").unwrap();

            if type_or_size == "dir" {
                current.children.push(Directory { 
                    name: name.to_string(), 
                    files: HashMap::new(),
                    children: vec![] 
                });
            } else {
                let size: u32 = type_or_size.parse().unwrap();
                current.files.insert(name.to_string(), size);
            }
        }
    }

    println!("{:?}", root);
	let answer1 = part1(&root);
	let answer2 = part2(&root);

	println!("Answer 1: {}", answer1);
	println!("Answer 2: {}", answer2);

    Ok(())
}
